src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -496,96 +508,88 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
                 }
                 pMessage->flags = _socketFlags;
             }
-        }
 
-        private unsafe IntPtr PtrWSAMessageBuffer
-        {
-            get
-            {
-                Debug.Assert(_wsaMessageBuffer != null);
-                Debug.Assert(_wsaMessageBuffer.Length == sizeof(Interop.Winsock.WSAMsg));
-                Debug.Assert(_wsaMessageBufferGCHandle.IsAllocated);
-                Debug.Assert(_wsaMessageBufferGCHandle.Target == _wsaMessageBuffer);
-                fixed (void* ptrWSAMessageBuffer = &_wsaMessageBuffer[0])
-                {
-                    return (IntPtr)ptrWSAMessageBuffer;
-                }
-            }
-        }
-
-        internal unsafe SocketError DoOperationReceiveMessageFrom(Socket socket, SafeCloseSocket handle)
-        {
             SocketError socketError = SocketError.Success;
             NativeOverlapped* overlapped = AllocateNativeOverlapped();
             try
             {
-                int bytesTransferred;
-
                 socketError = socket.WSARecvMsg(
                     handle,
-                    PtrWSAMessageBuffer,
-                    out bytesTransferred,
+                    Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBuffer, 0),
+                    out int bytesTransferred,
                     overlapped,
                     IntPtr.Zero);
 
-                socketError = ProcessIOCPResult(socketError == SocketError.Success, bytesTransferred);
+                socketError = ProcessIOCPResultWithSingleBufferHandle(socketError, bytesTransferred);
```

Comment:
Isn't this path handling both single-buffer and multi-buffer?

Reply:
It is.  If it's the single buffer case, then the handle state will already be Set.  If it's the multi buffer case, then the handle state will be None.  Either way, the `_singleBufferHandleState == SingleBufferHandleState.InProcess` check will be skipped.

Is the concern with the name?  Couldn't think of anything better, but certainly open to suggestions.

Reply:
Maybe just add a comment that calling it in the multi-buffer case is fine.